src/conv/invokers/mlir_impl_gemm.cpp

```diff
@@ -0,0 +1,134 @@
+// embedded into the permutation of strides.
+// - For NCHW, sizes = {N, C, H, W}; strides = {C*H*W, H*W, W, 1}
+// - For NHWC, sizes = {N, H, W, C}; strides = {C*H*W, W*C, C, 1}
+
+// In MIOpen however, size and strides are not aligned. Permutation of the strides are used to
+// infer actual layout
+// - For NCHW, sizes = {N, C, H, W}; strides = {C*H*W, H*W, W, 1}
+// - For NHWC, sizes = {N, C, H, W}; strides = {C*H*W, 1, W*C, C}
+auto permuteDimsStrides(const std::vector<size_t>& dims, const std::vector<size_t>& strides)
+{
+    auto sorted_dims    = dims;
+    auto sorted_strides = strides;
+    auto p              = TensorDescriptor::sort_permutation(strides, std::greater<>{});
+    std::transform(p.begin(), p.end(), sorted_dims.begin(), [&](auto i) { return dims[i]; });
+    std::transform(p.begin(), p.end(), sorted_strides.begin(), [&](auto i) { return strides[i]; });
+    return std::make_tuple(sorted_dims, sorted_strides);
+};
+
+void permuteDimStridesAllDir(const conv::ProblemDescription& conv_problem,
+                             std::vector<size_t>& in_dims,
+                             std::vector<size_t>& in_strides,
+                             std::vector<size_t>& weights_dims,
+                             std::vector<size_t>& weights_strides,
+                             std::vector<size_t>& out_dims,
+                             std::vector<size_t>& out_strides)
+{
+    const TensorDescriptor& in = conv_problem.GetIn();
+    std::make_tuple(in_dims, in_strides) = permuteDimsStrides(in.GetLengths(), in.GetStrides());
+
+    const TensorDescriptor& weights = conv_problem.GetWeights();
+    std::make_tuple(weights_dims, weights_strides) =
+        permuteDimsStrides(weights.GetLengths(), weights.GetStrides());
+
+    const TensorDescriptor& out = conv_problem.GetOut();
+    std::make_tuple(out_dims, out_strides) = permuteDimsStrides(out.GetLengths(), out.GetStrides());
+}
+
+MlirConvArgs MakeMlirConvArgs(ConstData_t in,
+                              const std::vector<size_t>& in_dims,
+                              const std::vector<size_t>& in_strides,
+                              ConstData_t w,
+                              const std::vector<size_t>& weights_dims,
+                              const std::vector<size_t>& weights_strides,
+                              ConstData_t out,
+                              const std::vector<size_t>& out_dims,
+                              const std::vector<size_t>& out_strides)
+{
+    auto cpyToMemRef = [](const void* ptr,
+                          const std::vector<size_t>& dims,
+                          const std::vector<size_t>& strides,
+                          MemRef4DGeneric& target) {
+        target.basePtr = const_cast<void*>(ptr); // NOLINT (cppcoreguidelines-pro-type-const-cast)
+        target.data    = const_cast<void*>(ptr); // NOLINT (cppcoreguidelines-pro-type-const-cast)
```

Comment:
I can understand `const_cast` is bad in general, though at this specific case it is harmless.

Will address this in next commit.

Reply:
I agree with @atamazov, `const_cast` is a bad idea.

Reply:
To me, it is just not cost effective to add a const to duplicate the data structure that the device interface doesn't even care about.

Refer to [here](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/pull/136/files). 

For every `MemRef4D`, I will have to duplicate the exported type for a const and regular pointer type just to get rid of this `const_cast` here.

Reply:
BTW @pfultz2 In the latest commit 29d5475, I have to make the `const_cast` used in combination with `reinterpret_cast`. Both are generally bad ideas in host code, but device function just only need a pointer in that offset.

Reply:
> For every MemRef4D, I will have to duplicate the exported type for a const and regular pointer type just to get rid of this const_cast here.

You should just have two different function to set the pointer as either `const` or not, and then the `const_cast` can happen in llvm if it necessary. This really should not be pushed on to clients.

Reply:
@pfultz2 LLVM only populate a device code object. `const` is a C/C++ semantic but the same concept does not exist in device or LLVM level code. In other words, if the MLIR wise populated device code doesn't care about const, and doesn't even have this abstraction, it isn't fair for the library to force this `const` as well.

Reply:
> To me, it is just not cost effective to add a const to duplicate the data structure that the device interface doesn't even care about.

This is important for OpenCL backend.

> const is a C/C++ semantic but the same concept does not exist in device or LLVM level code. 

[off-topic] Of course this semantic is important for device code, and in llvm too. 

> it isn't fair for the library to force this const as well.

[off-topic] But we didn’t promise _fair play_ :smile:

Reply:
Will address according to @pfultz2 [comment in this PR](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/pull/136). It looks like a much more robust approach compared to exposing the structure in the interface. With that, I can create two interface function that help to address MIOpen's concern here.

Notice that both function return the similar `MiirStridedMemRef` because const-ness does not impact data layout when passing arguments to device.

```cpp
MiirStridedMemRef MiirCreateStridedMemRef(void* data,
                                          int64_t offset,
                                          int64_t * sizes,
                                          int64_t * strides,
                                          int64_t ndims); 
{
    return new StridedMemRefType<void*, ndims>(data, data, offset, sizes, strides);
}

MiirStridedMemRef MiirCreateConstStridedMemRef(const void* data,
                                               int64_t offset,
                                               int64_t * sizes,
                                               int64_t * strides,
                                               int64_t ndims); 
{
    return new StridedMemRefType<const void*, ndims>(data, data, offset, sizes, strides);
}
```

Reply:
@jerryyin These functions create `new` objects. Who will `delete` them?

Reply:
@atamazov Paul already give a solution in that comments:

`void MiirDestroyStridedMemRef(MiirStridedMemRef x);`

We can create a shared ptr (with this custom deleter) to do this. Is that a infrastrcture already exist in MIOpen, like `MIOPEN_MANAGE_PTR`, but for `shared_ptr`?

Reply:
See https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/pull/136#issuecomment-820661976

Reply:
> Is that a infrastrcture already exist in MIOpen, like MIOPEN_MANAGE_PTR, but for shared_ptr?

You can construct it first as a unique_ptr using `MIOPEN_MANAGE_PTR` then you can pass it to a shared_ptr. A shared_ptr uses a type erased deleter so the type of a shared_ptr doesn't change based on the deleter. I suggest you write a simple function to construct it as a shared_ptr, so this can all be done in one place.

Reply:
After I figured that MLIR side cannot change the device ABI, I decide to not pursue further with this opaque pointer idea. Sorry for the confusions I may have caused on this thread.

With the definition of the structure on the interface, I'd still prefer to keep the `const_cast`on MIOpen end.